src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -488,18 +489,26 @@ static void gavl_tree_delete_internal(gavl_tree_s * tree_ptr, gavl_tree_node_s *
         } else {
             dnode->right = NULL;
         }
-        val = (void *) dnode->val;
-        /* move inorder_node's buffer object to dnode and then free inorder_node */
-        dnode->addr = inorder_node->addr;
-        dnode->len = inorder_node->len;
-        dnode->val = inorder_node->val;
-        dnode = inorder_node;
-    }
 
-    /* free object using user-provided free function */
-    if (tree_ptr->gavl_free_fn)
-        tree_ptr->gavl_free_fn(val);
-    MPL_free(dnode);
+        /* move inorder_node to dnode position and remove dnode from tree */
+        inorder_node->parent = dnode->parent;
```

Comment:
A side questions: have you thought about code without the parent pointer? Since you already have the stack, you always know your parent, right? The extra pointer means more code to maintain.

Reply:
parent pointer will be used in line 315-316
```
while (tree_ptr->root && tree_ptr->root->parent != NULL)
            tree_ptr->root = tree_ptr->root->parent;
```

Reply:
I see.

During  the  rebalance, you always know which node is under rotation, and you always can tell whether that node is root, right? Couldn't you simply update the root whenever that occurs?

Reply:
emm, good idea. I can do that.

Reply:
I feel this change is unrelated to this PR, I will create another PR to change it.